# Exercise 9

## Mashable news stories analysis

Predicting if a news story is going to be popular


In [3]:
import pandas as pd

url = 'https://raw.githubusercontent.com/albahnsen/PracticalMachineLearningClass/master/datasets/mashable.csv'
df = pd.read_csv(url, index_col=0)
df.head()

url  timedelta  \
0  http://mashable.com/2014/12/10/cia-torture-rep...       28.0   
1  http://mashable.com/2013/10/18/bitlock-kicksta...      447.0   
2  http://mashable.com/2013/07/24/google-glass-po...      533.0   
3  http://mashable.com/2013/11/21/these-are-the-m...      413.0   
4  http://mashable.com/2014/02/11/parking-ticket-...      331.0   

   n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  \
0             9.0             188.0         0.732620               1.0   
1             7.0             297.0         0.653199               1.0   
2            11.0             181.0         0.660377               1.0   
3            12.0             781.0         0.497409               1.0   
4             8.0             177.0         0.685714               1.0   

   n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs   ...     \
0                  0.844262        5.0             1.0       1.0   ...      
1                  0.815789        9.0             4.0       1.0   ...      
2                  0.775701        4.0             3.0       1.0   ...      
3                  0.677350       10.0             3.0       1.0   ...      
4                  0.830357        3.0             2.0       1.0   ...      

   min_positive_polarity  max_positive_polarity  avg_negative_polarity  \
0               0.200000                   0.80              -0.487500   
1               0.160000                   0.50              -0.135340   
2               0.136364                   1.00               0.000000   
3               0.100000                   1.00              -0.195701   
4               0.100000                   0.55              -0.175000   

   min_negative_polarity  max_negative_polarity  title_subjectivity  \
0                  -0.60              -0.250000                 0.9   
1                  -0.40              -0.050000                 0.1   
2                   0.00               0.000000                 0.3   
3                  -0.40              -0.071429                 0.0   
4                  -0.25              -0.100000                 0.0   

   title_sentiment_polarity  abs_title_subjectivity  \
0                       0.8                     0.4   
1                      -0.1                     0.4   
2                       1.0                     0.2   
3                       0.0                     0.5   
4                       0.0                     0.5   

   abs_title_sentiment_polarity  Popular  
0                           0.8        1  
1                           0.1        0  
2                           1.0        0  
3                           0.0        0  
4                           0.0        0  

[5 rows x 61 columns]

In [4]:
df.shape

(6000, 61)

In [5]:
X = df.drop(['url', 'Popular'], axis=1)
y = df['Popular']

In [6]:
y.mean()

0.5

In [7]:
# train/test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [8]:
y_test.shape

(1500,)

# Exercise 9.1

Estimate a Decision Tree Classifier and a Logistic Regresion

Evaluate using the following metrics:
* Accuracy
* F1-Score

In [9]:
from sklearn.tree import DecisionTreeClassifier
treeclf = DecisionTreeClassifier(random_state=1)
treeclf.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=1,
            splitter='best')

In [10]:
from sklearn.metrics import accuracy_score
treeclf.score(X_test, y_test)
y_pred = treeclf.predict(X_test)
score=treeclf.score(X_test, y_test)
print('Score: \n',score)
accuracy = accuracy_score(y_test, y_pred, normalize=False)
print('Casos clasificados con precision: ' + str(accuracy))

Score: 
 0.5426666666666666
Casos clasificados con precision: 814


In [11]:
from sklearn.metrics import f1_score
y_pred = treeclf.predict(X_test) 
print('f1 score(default): \n',  f1_score(y_test, y_pred))


f1 score(default): 
 0.5402144772117962


In [12]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(random_state=0).fit(X_train, y_train)
print('Score: \n',score)
y_pred = clf.predict(X_test) 
print('f1 score(default): \n',  f1_score(y_test, y_pred))


Score: 
 0.5426666666666666
f1 score(default): 
 0.6104529616724739


#### Mientras que para el árbol de decisión el F1 score es igual a 0.54 para la regresión logistica es 0.61, que interpretamos como el grado ponderado de clasificaciones correctas que hace respectivamente cada uno de los modelos, si apreciamos el score este no es ponderado y el resultado es el mismo para ambos modelos

# Exercise 9.2

Estimate 300 bagged samples

Estimate the following set of classifiers:

* 100 Decision Trees where max_depth=None
* 100 Decision Trees where max_depth=2
* 100 Logistic Regressions

In [13]:
print(X_test.shape[0])

1500


In [14]:
import numpy as np
# set a seed for reproducibility

np.random.seed(123)
n_samples = X_train.shape[0]
n_B = 100
# create ten bootstrap samples (will be used to select rows from the DataFrame)
samples_1 = [np.random.choice(a=n_samples, size=n_samples, replace=True) for _ in range(1, n_B +1 )]

In [15]:
np.random.seed(123)
n_samples = X_train.shape[0]
n_B = 100
# create ten bootstrap samples (will be used to select rows from the DataFrame)
samples_2 = [np.random.choice(a=n_samples, size=n_samples, replace=True) for _ in range(1, n_B +1 )]

In [16]:
np.random.seed(123)
n_samples = X_train.shape[0]
n_B = 100
# create ten bootstrap samples (will be used to select rows from the DataFrame)
samples_3 = [np.random.choice(a=n_samples, size=n_samples, replace=True) for _ in range(1, n_B +1 )]

In [17]:
# show the rows for the first decision tree
data1=X_train.iloc[samples_1[0], :]
data1.head(5)

timedelta  n_tokens_title  n_tokens_content  n_unique_tokens  \
2239      142.0            11.0             135.0         0.631579   
4838      349.0            10.0            1411.0         0.431259   
1047      731.0            12.0             238.0         0.621277   
373       583.0             8.0             253.0         0.661355   
2595      659.0            15.0             458.0         0.512088   

      n_non_stop_words  n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  \
2239               1.0                  0.690000       11.0             4.0   
4838               1.0                  0.630996        7.0             6.0   
1047               1.0                  0.823077        3.0             1.0   
373                1.0                  0.846667        7.0             0.0   
2595               1.0                  0.662879        5.0             3.0   

      num_imgs  num_videos              ...               \
2239       8.0         0.0              ...                
4838       4.0         1.0              ...                
1047       1.0         0.0              ...                
373        0.0         1.0              ...                
2595       1.0         0.0              ...                

      avg_positive_polarity  min_positive_polarity  max_positive_polarity  \
2239               0.354762               0.285714                    0.5   
4838               0.438312               0.050000                    1.0   
1047               0.291259               0.100000                    0.6   
373                0.432149               0.050000                    1.0   
2595               0.368586               0.136364                    0.8   

      avg_negative_polarity  min_negative_polarity  max_negative_polarity  \
2239              -0.308333                 -0.625              -0.100000   
4838              -0.322500                 -0.800              -0.100000   
1047               0.000000                  0.000               0.000000   
373               -0.247194                 -0.500              -0.071429   
2595              -0.219167                 -0.500              -0.125000   

      title_subjectivity  title_sentiment_polarity  abs_title_subjectivity  \
2239            0.333333                  0.250000                0.166667   
4838            1.000000                  0.100000                0.500000   
1047            0.000000                  0.000000                0.500000   
373             0.400000                 -0.400000                0.100000   
2595            0.454545                  0.136364                0.045455   

      abs_title_sentiment_polarity  
2239                      0.250000  
4838                      0.100000  
1047                      0.000000  
373                       0.400000  
2595                      0.136364  

[5 rows x 59 columns]

In [18]:
data1.shape

(4500, 59)

In [20]:
# Estimate 100 Decision Trees where max_depth=None

from sklearn.tree import DecisionTreeClassifier

# grow each tree deep
treeclas = DecisionTreeClassifier(max_depth=None, random_state=123)

# DataFrame for storing predicted price from each tree
y_pred_1 = pd.DataFrame(index=y_test.index, columns=[list(range(n_B))])

# grow one tree for each bootstrap sample and make predictions on testing data
for i, sample in enumerate(samples_1):
    #print(i)
    #print(sample)
    X1_train = X_train.iloc[sample, :]
    y1_train = y_train.iloc[sample, ]
    treeclas.fit(X1_train, y1_train)
    y_pred_1[i] = treeclas.predict(X_test)

In [21]:
y_pred_1

0  1  2  3  4  5  6  7  8  9  ... 90 91 92 93 94 95 96 97 98 99
1483  1  1  0  0  1  1  0  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
2185  1  1  1  1  0  1  0  1  1  0 ...  1  0  1  1  0  1  1  1  1  1
2520  1  0  0  0  1  1  1  1  0  0 ...  1  1  1  0  1  0  1  0  1  1
3721  1  0  1  0  1  1  1  0  1  1 ...  1  1  1  1  0  1  0  1  1  1
3727  0  0  0  0  1  0  1  0  0  0 ...  0  0  1  1  0  0  0  0  1  0
4524  0  0  0  1  0  0  1  1  0  0 ...  1  1  0  0  1  1  0  1  1  0
234   0  0  0  0  0  0  0  0  0  0 ...  0  1  0  0  1  0  0  1  1  0
4735  1  1  1  1  0  1  1  1  0  1 ...  1  1  1  0  1  0  0  1  1  0
5839  0  1  0  0  1  1  1  0  1  0 ...  1  0  1  1  0  1  0  0  1  0
2939  0  0  1  1  1  1  0  1  1  0 ...  1  0  0  1  1  1  1  0  1  1
3053  1  1  1  0  0  0  0  1  1  1 ...  1  0  0  1  1  1  1  1  1  0
867   1  1  1  1  1  1  1  0  1  0 ...  0  0  1  1  1  0  0  0  0  1
276   1  1  0  0  1  1  1  0  1  0 ...  1  1  1  1  1  0  1  1  0  1
5798  0  1  0  0  1  0  0  0  0  0 ...  0  0  1  0  1  0  0  0  0  0
3512  1  1  1  0  0  0  0  0  1  0 ...  1  1  1  1  1  1  0  1  1  1
5789  1  0  0  1  1  1  0  0  0  0 ...  1  0  0  0  0  0  0  0  0  0
3198  1  0  0  1  1  1  1  1  0  1 ...  1  0  0  1  1  1  1  1  0  0
5908  1  0  1  1  1  1  1  1  1  1 ...  0  1  1  1  1  1  1  1  1  0
5718  0  1  1  1  0  0  0  0  1  0 ...  0  0  0  0  1  1  1  0  0  0
2687  0  0  0  0  1  0  0  0  0  0 ...  0  0  1  0  0  0  0  0  0  1
250   0  1  1  1  1  1  0  0  0  0 ...  1  0  1  1  1  1  1  0  1  1
5461  0  0  0  1  1  0  1  1  1  1 ...  1  1  0  1  1  0  1  0  0  1
3052  1  0  0  0  0  0  0  0  0  0 ...  0  0  1  0  0  0  1  0  0  0
2711  1  1  1  1  1  1  1  0  1  0 ...  1  1  0  1  0  1  1  0  1  1
3771  1  1  0  0  0  1  1  0  0  0 ...  0  0  1  0  1  0  1  0  0  0
5988  0  0  0  0  0  0  1  0  1  0 ...  0  0  1  0  0  1  1  1  0  0
5645  0  1  0  1  1  1  0  0  1  1 ...  0  0  1  0  1  0  1  1  1  0
4617  0  0  1  0  0  1  0  0  1  0 ...  0  0  0  1  1  0  0  1  1  1
368   1  0  1  1  0  1  1  0  1  1 ...  1  0  1  1  0  1  1  1  1  1
4916  1  1  1  1  1  1  1  1  1  0 ...  1  0  0  1  1  0  0  1  1  0
...  .. .. .. .. .. .. .. .. .. .. ... .. .. .. .. .. .. .. .. .. ..
19    1  1  0  0  1  0  1  0  0  1 ...  1  1  0  0  1  1  0  1  0  1
4088  1  0  1  1  0  1  1  0  0  0 ...  1  1  1  1  1  1  1  1  1  1
1395  1  0  0  1  1  1  1  1  1  1 ...  1  0  0  1  1  1  1  1  1  1
2576  0  1  0  0  1  1  0  1  0  1 ...  1  1  1  0  1  1  1  1  0  0
2878  1  0  0  1  0  1  0  1  1  0 ...  0  1  1  1  1  0  1  1  1  0
3147  1  1  1  1  0  0  0  1  1  0 ...  0  1  0  1  0  1  1  1  1  1
4513  1  1  1  0  0  0  0  1  0  0 ...  1  0  1  0  1  1  1  1  1  0
1620  1  1  1  0  0  0  1  1  1  1 ...  1  0  1  0  0  1  1  1  1  0
4522  1  0  1  1  1  1  0  1  1  0 ...  1  0  1  0  0  1  1  1  0  1
5485  0  1  1  1  1  1  1  0  0  1 ...  0  0  1  1  1  0  1  1  1  1
2394  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  1  1  0  0  0  0
1691  1  1  1  1  0  1  1  0  0  1 ...  1  1  0  0  1  1  1  1  0  1
4949  0  0  0  0  0  1  0  1  1  0 ...  0  1  0  1  1  1  1  1  1  0
2805  1  1  0  1  0  1  1  1  0  1 ...  1  1  0  0  0  1  1  0  0  0
494   0  1  0  1  1  1  0  0  1  1 ...  1  1  1  1  1  0  1  0  1  1
5287  0  1  0  0  0  0  1  0  0  1 ...  0  0  0  0  1  0  1  0  0  0
3316  0  1  0  0  0  0  1  0  0  0 ...  1  1  1  1  1  1  0  1  0  0
3176  1  0  1  0  1  1  1  1  1  1 ...  1  1  1  1  0  0  0  1  1  0
2372  1  1  1  1  0  1  0  0  1  0 ...  1  0  1  0  0  0  1  0  1  1
5947  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  0  1  1  0  1  0  1
5782  1  1  1  0  0  0  0  1  0  0 ...  0  1  0  1  0  0  0  1  0  0
1814  1  0  0  1  1  0  0  0  1  0 ...  0  0  0  1  1  0  0  0  0  1
4704  0  0  0  0  1  0  0  0  0  0 ...  0  0  1  0  0  0  0  0  0  0
5333  0  0  1  0  1  0  0  1  1  0 ...  0  1  1  0  0  1  0  1  0  0
946   1  0  0  0  1  0  0  1  0  1 ...  0  0  1  1  1  0  1  1  1  0
3077  1  1  0  0  1  1  0  1  0  0 ...  0  1  1  0  0  1  0  0  0  0
516

In [22]:
# Estimate 100 Decision Trees where max_depth=None

from sklearn.tree import DecisionTreeClassifier

# grow each tree deep
treeclas = DecisionTreeClassifier(max_depth=2, random_state=123)

# DataFrame for storing predicted price from each tree
y_pred_2 = pd.DataFrame(index=y_test.index, columns=[list(range(n_B))])

# grow one tree for each bootstrap sample and make predictions on testing data
for i, sample in enumerate(samples_2):
    #print(i)
    #print(sample)
    X1_train = X_train.iloc[sample, :]
    y1_train = y_train.iloc[sample, ]
    treeclas.fit(X1_train, y1_train)
    y_pred_2[i] = treeclas.predict(X_test)

In [23]:
y_pred_2

0  1  2  3  4  5  6  7  8  9  ... 90 91 92 93 94 95 96 97 98 99
1483  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
2185  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
2520  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
3721  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
3727  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
4524  1  1  1  1  0  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
234   0  0  0  0  0  0  0  0  0  0 ...  0  0  1  0  0  0  0  0  0  0
4735  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  1  0  0  0  0
5839  0  0  0  0  0  0  0  0  0  0 ...  0  0  1  0  0  0  0  0  0  0
2939  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
3053  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
867   1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
276   1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5798  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
3512  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5789  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
3198  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5908  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5718  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  1  0  0  0  0
2687  0  0  0  0  0  0  0  0  0  0 ...  0  0  1  0  0  0  0  0  0  0
250   1  1  1  1  0  0  1  1  1  1 ...  0  1  1  1  1  1  1  0  1  1
5461  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
3052  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
2711  1  1  1  1  0  0  1  1  1  1 ...  1  1  1  1  1  1  1  0  1  1
3771  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
5988  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  1  0  0  0  0
5645  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
4617  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
368   1  1  1  1  0  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
4916  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
...  .. .. .. .. .. .. .. .. .. .. ... .. .. .. .. .. .. .. .. .. ..
19    0  0  0  0  0  0  0  1  0  0 ...  0  0  0  0  0  0  0  0  0  0
4088  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
1395  1  1  1  1  0  0  1  1  1  1 ...  1  1  1  1  1  1  1  0  1  1
2576  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  1  0  0  0  0
2878  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
3147  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
4513  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
1620  1  1  1  1  0  0  1  1  1  1 ...  1  1  1  1  1  1  1  0  1  1
4522  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5485  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
2394  0  0  0  0  0  0  0  0  0  0 ...  0  0  1  0  0  0  0  0  0  0
1691  0  0  0  0  0  0  0  0  0  0 ...  1  0  0  0  1  1  0  0  0  0
4949  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
2805  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
494   1  1  1  1  0  0  1  1  1  1 ...  1  1  1  1  1  1  1  0  1  1
5287  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
3316  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
3176  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
2372  1  1  1  1  0  0  1  1  1  1 ...  1  1  1  1  1  1  1  0  1  1
5947  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5782  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
1814  0  0  0  0  0  0  0  0  0  0 ...  0  0  1  0  0  0  0  0  0  0
4704  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
5333  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  1  0  0  0  0
946   1  1  1  1  0  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
3077  1  1  1  1  0  0  1  1  1  1 ...  1  1  1  1  1  1  1  0  1  1
516

In [24]:
# fit a logistic regression model and store the class predictions
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression(solver='liblinear',C=1e9)

# DataFrame for storing predicted price from each tree
y_pred_3 = pd.DataFrame(index=y_test.index, columns=[list(range(n_B))])

# grow one tree for each bootstrap sample and make predictions on testing data
for i, sample in enumerate(samples_3):
    #print(i)
    #print(sample)
    X1_train = X_train.iloc[sample, :]
    y1_train = y_train.iloc[sample, ]
    logreg.fit(X1_train, y1_train)
    y_pred_3[i] = logreg.predict(X_test)


In [25]:
y_pred_3

0  1  2  3  4  5  6  7  8  9  ... 90 91 92 93 94 95 96 97 98 99
1483  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
2185  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
2520  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
3721  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
3727  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
4524  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
234   0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
4735  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
5839  1  1  0  0  1  1  1  1  1  1 ...  1  0  1  1  1  1  0  1  0  0
2939  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
3053  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
867   1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
276   1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5798  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
3512  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5789  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
3198  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5908  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5718  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
2687  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
250   0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
5461  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
3052  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
2711  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
3771  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
5988  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
5645  0  0  0  0  0  0  0  1  0  0 ...  0  0  0  0  0  0  0  0  0  0
4617  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
368   1  1  1  1  1  1  1  1  1  1 ...  0  1  1  1  1  1  1  1  1  1
4916  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
...  .. .. .. .. .. .. .. .. .. .. ... .. .. .. .. .. .. .. .. .. ..
19    0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
4088  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
1395  1  1  1  1  0  1  0  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
2576  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
2878  1  1  1  0  0  1  0  1  1  1 ...  0  0  1  0  1  1  0  0  1  0
3147  1  1  0  0  0  1  0  1  0  0 ...  0  0  0  0  1  1  0  0  1  0
4513  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
1620  0  0  0  0  0  0  0  0  0  0 ...  0  1  0  0  0  0  0  0  0  0
4522  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5485  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
2394  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
1691  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
4949  0  1  1  0  1  1  1  1  1  1 ...  0  0  1  1  0  0  0  0  0  0
2805  0  0  0  0  0  1  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
494   1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5287  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
3316  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
3176  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
2372  1  1  1  1  1  1  0  0  1  1 ...  1  1  1  1  1  1  1  1  1  1
5947  1  1  1  1  1  1  1  1  1  1 ...  1  1  1  1  1  1  1  1  1  1
5782  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
1814  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
4704  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
5333  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
946   0  1  0  0  0  0  0  0  1  0 ...  0  1  1  0  0  0  0  1  1  1
3077  0  0  0  0  0  0  0  0  0  0 ...  0  0  0  0  0  0  0  0  0  0
516

#### Se crean las 100 muestras para cada uno de los modelos

# Exercise 9.3

Ensemble using majority voting

Evaluate using the following metrics:
* Accuracy
* F1-Score

In [26]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingRegressor
from sklearn.ensemble import BaggingClassifier
from sklearn.metrics import accuracy_score
from sklearn import metrics

In [27]:
clf = BaggingClassifier(base_estimator=DecisionTreeClassifier(max_depth=None, random_state=123), 
                        n_estimators=100, bootstrap=True, random_state=42, n_jobs=-1, oob_score=True)

clf.fit(X_train, y_train)
y_pred_1 = clf.predict(X_test)
metrics.f1_score(y_pred_1, y_test), metrics.accuracy_score(y_pred_1, y_test)

(0.6456278763971072, 0.6406666666666667)

In [28]:
y_pred_1

array([1, 1, 1, ..., 1, 0, 0], dtype=int64)

#### Utilzando Bagging podemos ver que el F1 Score pasa  de 0.54 a 0.64, lo cual representa un crecimiento significativo en la capacidad de clasificar bien el modelo

In [29]:
clf = BaggingClassifier(base_estimator=DecisionTreeClassifier(max_depth=2, random_state=123), 
                        n_estimators=100, bootstrap=True, random_state=42, n_jobs=-1, oob_score=True)

clf.fit(X_train, y_train)
y_pred_2 = clf.predict(X_test)
metrics.f1_score(y_pred_2, y_test), metrics.accuracy_score(y_pred_2, y_test)

(0.6439545758183033, 0.6446666666666667)

In [30]:
y_pred_2

array([1, 1, 1, ..., 1, 0, 0], dtype=int64)

In [31]:
clf = BaggingClassifier(base_estimator=LogisticRegression(solver='liblinear',C=1e9, random_state=123), 
                        n_estimators=100, bootstrap=True, random_state=42, n_jobs=-1, oob_score=True)

clf.fit(X_train, y_train)
y_pred_3 = clf.predict(X_test)
metrics.f1_score(y_pred_3, y_test), metrics.accuracy_score(y_pred_3, y_test)

(0.6065232477446219, 0.622)

#### Utilzando Bagging podemos ver que el F1 Score pasa  de 0.61 a 0.62, lo cual representa un crecimiento menor en la capacidad de clasificar bien el modelo, pero al final un resultado mejor que con el modelo original

# Exercise 9.4

Estimate te probability as %models that predict positive

Modify the probability threshold and select the one that maximizes the F1-Score

In [32]:
prob_mod1=sum(y_pred_1)/y_pred_1.shape
print(prob_mod1)
prob_mod2=sum(y_pred_2)/y_pred_2.shape
print(prob_mod2)
prob_mod3=sum(y_pred_3)/y_pred_3.shape
print(prob_mod3)


[0.51]
[0.494]
[0.45666667]


In [36]:
import numpy as np
import warnings
warnings.filterwarnings("ignore")
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score

clf = BaggingClassifier(base_estimator=DecisionTreeClassifier(max_depth=2, random_state=123), 
                        n_estimators=100, bootstrap=True, random_state=42, n_jobs=-1, oob_score=True)
fs=[]
rango=np.arange(0.3,1,0.1)
for j in rango:
    clf.fit(X_train, y_train)
    y_pred = (clf.predict_proba(X_test)[:,1]>=j).astype(int)
    fs.append([j,round(f1_score(y_pred,y_test),2),round(clf.score(X_test,y_test),2)])
fs=pd.DataFrame(fs, columns=['prob','F1_score','tree_score'])
fs.loc[fs['F1_score'].idxmax()]

prob          0.40
F1_score      0.70
tree_score    0.64
Name: 1, dtype: float64

In [37]:
y_pred

array([0, 0, 0, ..., 0, 0, 0])

#### Ahora al variar los limites para la construcción de los casos positivos (valores 1) pódemos ver que con el 0.4 logramos el mayor % de casos exitosos de clasifiacción, logrando una clasificación correcta del 70%, esto es concordante con el teorema de condorcet ya que la probailidad de 1 es igual a 0.6 y por lo tanto el resultado combinado resulta superior y en este caso a 10 unidades. 

# Exercise 9.5

Ensemble using weighted voting using the oob_error

Evaluate using the following metrics:
* Accuracy
* F1-Score

In [60]:
models=[DecisionTreeClassifier(max_depth=None, random_state=123),DecisionTreeClassifier(max_depth=2, random_state=123),
        LogisticRegression(solver='liblinear',C=1e9, random_state=123),]
print(models)

y_pred_mean=[]
y_pred_df=[]
fs=[]
j=1
for model in list(models):
    clf = BaggingClassifier(base_estimator=model,n_estimators=100, bootstrap=True, random_state=42, n_jobs=-1, oob_score=True)
    clf.fit(X_train,y_train)
    errors=np.zeros(clf.n_estimators)
    y_pred_all_=np.zeros((X_test.shape[0],clf.n_estimators))
    for b in range(clf.n_estimators):
        oob_sample = ~clf.estimators_samples_[b]
        y_pred_=clf.estimators_[b].predict(pd.DataFrame(X_train).values[oob_sample])
        errors[b]=metrics.accuracy_score(y_pred_, y_train.values[oob_sample])
        y_pred_all_[:,b]=clf.estimators_[b].predict(X_test)
    alpha=(1-errors)/ (1-errors).sum()
    y_pred=(np.sum(y_pred_all_*alpha, axis=1)>=0.5).astype(np.int)
    y_pred_df.append([y_pred])
    y_pred_mean.append([b, y_pred.mean()])
    fs.append([j,round(f1_score(y_pred,y_test),3),round(clf.score(X_test,y_test),3)])
    j=j+1
y_pred_mean=pd.DataFrame(y_pred_mean)
y_pred_mean[1].tolist()
fs=pd.DataFrame(fs, columns=['prob','F1_score','tree_score'])

[DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=123,
            splitter='best'), DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=2,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=123,
            splitter='best'), LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
          fit_intercept=True, intercept_scaling=1, max_iter=100,
          multi_class='ovr', n_jobs=1, penalty='l2', random_state=123,
          solver='liblinear', tol=0.0001, verbose=0, warm_start=False)]


In [61]:
print(fs)

   prob  F1_score  tree_score
0     1     0.650       0.641
1     2     0.645       0.645
2     3     0.609       0.622


In [64]:
print(y_pred_df[1])

[array([1, 1, 1, ..., 1, 0, 0])]


#### volviendo al criterio de 1 de un 50%, podemos encontrar que al incluir el concepto de OOB para mejorar las predcciones el F1 score crece en el caso del modelo de árboles de decisión a 64.5% y el de regresión logística a 65%

# Exercise 9.6

Estimate te probability of the weighted voting

Modify the probability threshold and select the one that maximizes the F1-Score

In [132]:
y_pred_0= np.asarray(y_pred_df[0])
prob_mod1=np.sum(y_pred_0)/y_pred_0.shape[1]
print(prob_mod1)

0.496


#### Aunque el 50% de los registros son 1, el F1 score clasifica correctamente un 65% (Arbol de clasificacion, profundidad= no especificada)

In [133]:
y_pred_1= np.asarray(y_pred_df[1])
prob_mod2=np.sum(y_pred_1)/y_pred_1.shape[1]
print(prob_mod2)

0.472


#### Aunque el 47.2% de los registros son 1, el F1 score clasifica correctamente un 64.5% (Arbol de clasificacion, profundidad=2)

In [134]:
y_pred_2= np.asarray(y_pred_df[1])
prob_mod3=np.sum(y_pred_2)/y_pred_2.shape[1]
print(prob_mod3)

0.472


#### Aunque el 47.2% de los registros son 1, el F1 score clasifica correctamente un 60.9% (modelo de regresion logistica)

In [135]:
models=[DecisionTreeClassifier(max_depth=None, random_state=123),DecisionTreeClassifier(max_depth=2, random_state=123),
        LogisticRegression(solver='liblinear',C=1e9, random_state=123)]
print(models)
    
y_pred_mean=[]
y_pred_df=[]
fs=[]
fk=[]
j=1
for model in list(models):
    clf = BaggingClassifier(base_estimator=model,n_estimators=100, bootstrap=True, random_state=42, n_jobs=-1, oob_score=True)
    
    ft=[]
    rango=np.arange(0.3,1,0.1)
    for t in rango:
        clf.fit(X_train, y_train)
        errors=np.zeros(clf.n_estimators)
        y_pred_all_=np.zeros((X_test.shape[0],clf.n_estimators))
        for b in range(clf.n_estimators):
            oob_sample = ~clf.estimators_samples_[b]
            y_pred_=clf.estimators_[b].predict(pd.DataFrame(X_train).values[oob_sample])
            errors[b]=metrics.accuracy_score(y_pred_, y_train.values[oob_sample])
            y_pred_all_[:,b]=clf.estimators_[b].predict(X_test)
        alpha=(1-errors)/ (1-errors).sum()
        y_pred=(np.sum(y_pred_all_*alpha, axis=1)>=t).astype(np.int)
        y_pred_df.append([y_pred])
        y_pred_mean.append([b, y_pred.mean()])
        ft.append([t,round(f1_score(y_pred,y_test),3),round(clf.score(X_test,y_test),3)])
        j=j+1
    ft=pd.DataFrame(ft, columns=['prob','F1_score','tree_score'])
    print(ft)
    fs=ft.loc[ft['F1_score'].idxmax()]
    print(fs)
    fk.append([fk,fs])
    
print(fk)

[DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=123,
            splitter='best'), DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=2,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=123,
            splitter='best'), LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
          fit_intercept=True, intercept_scaling=1, max_iter=100,
          multi_class='ovr', n_jobs=1, penalty='l2', random_state=123,
          solver='liblinear', tol=0.0001, verbose=0, warm_start=False)]
   prob  F1_score  tree_scor

In [136]:
print(fk)

[[[...], prob          0.400
F1_score      0.701
tree_score    0.641
Name: 1, dtype: float64], [[...], prob          0.300
F1_score      0.648
tree_score    0.645
Name: 0, dtype: float64], [[...], prob          0.300
F1_score      0.623
tree_score    0.622
Name: 0, dtype: float64]]


#### Al iterar sobre los limites de clasificacion de 1 y 0, para cada uno de los modelos encontramos que, para el modelo de arbol de clasificacion con profundidad indefinidad el maximo F1 score se alcanza con un limite de >=0.4 para definir el 1,  en el caso del arbol de decision con profundidad 2 este limite es de 0.3 y finalmente con el modelo de regresion logistica es de 0.3

# Exercise 9.7

Estimate a logistic regression using as input the estimated classifiers

Modify the probability threshold such that maximizes the F1-Score

In [137]:
models=[LogisticRegression(solver='liblinear',C=1e9, random_state=123)]
print(models)
    
y_pred_mean=[]
y_pred_df=[]
fs=[]
fk=[]
j=1
for model in list(models):
    clf = BaggingClassifier(base_estimator=model,n_estimators=100, bootstrap=True, random_state=42, n_jobs=-1, oob_score=True)
    
    ft=[]
    rango=np.arange(0.3,1,0.1)
    for t in rango:
        clf.fit(X_train, y_train)
        errors=np.zeros(clf.n_estimators)
        y_pred_all_=np.zeros((X_test.shape[0],clf.n_estimators))
        for b in range(clf.n_estimators):
            oob_sample = ~clf.estimators_samples_[b]
            y_pred_=clf.estimators_[b].predict(pd.DataFrame(X_train).values[oob_sample])
            errors[b]=metrics.accuracy_score(y_pred_, y_train.values[oob_sample])
            y_pred_all_[:,b]=clf.estimators_[b].predict(X_test)
        alpha=(1-errors)/ (1-errors).sum()
        y_pred=(np.sum(y_pred_all_*alpha, axis=1)>=t).astype(np.int)
        y_pred_df.append([y_pred])
        y_pred_mean.append([b, y_pred.mean()])
        ft.append([t,round(f1_score(y_pred,y_test),3),round(clf.score(X_test,y_test),3)])
        j=j+1
    ft=pd.DataFrame(ft, columns=['prob','F1_score','tree_score'])
    print(ft)
    fs=ft.loc[ft['F1_score'].idxmax()]
    print(fs)
    fk.append([fk,fs]) 
print(fk)

[LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
          fit_intercept=True, intercept_scaling=1, max_iter=100,
          multi_class='ovr', n_jobs=1, penalty='l2', random_state=123,
          solver='liblinear', tol=0.0001, verbose=0, warm_start=False)]
   prob  F1_score  tree_score
0   0.3     0.623       0.622
1   0.4     0.616       0.622
2   0.5     0.609       0.622
3   0.6     0.602       0.622
4   0.7     0.589       0.622
5   0.8     0.576       0.622
6   0.9     0.566       0.622
prob          0.300
F1_score      0.623
tree_score    0.622
Name: 0, dtype: float64
[[[...], prob          0.300
F1_score      0.623
tree_score    0.622
Name: 0, dtype: float64]]


#### Al iterar sobre los limites de clasificacion de los casos positivos (1), para el modelo de Regresion Logistica el maximo F1 score se alcanza con un limite de >=0.3 para definir el 1, con este limite se alcanza un F1 score de 0.623